### PR TITLE
feat(chains): adding ElasticsearchDatabaseChain for interacting with analytics database

### PR DESCRIPTION
**Pull request by samber**
This pull request adds a ElasticsearchDatabaseChain chain for interacting with analytics database, in the manner of the SQLDatabaseChain.

It can be used in this way:

```py
from elasticsearch import Elasticsearch

from langchain.chains.elasticsearch_database import ElasticsearchDatabaseChain
from langchain.chat_models import ChatOpenAI

client = Elasticsearch("http://localhost:9200")

OPENAI_API_KEY = os.environ.get('OPENAI_API_KEY')
llm = ChatOpenAI(temperature=0, openai_api_key=OPENAI_API_KEY, model_name='gpt-3.5-turbo-16k')

db_chain = ElasticsearchDatabaseChain.from_llm(llm=llm, db=client, verbose=False, sample_documents_in_index_info=10)

result = db_chain.run(question)
print(result)
```

Maintainer: @samber
Twitter handler: samuelberthe

@baskaryan

### CODE REVIEW

File: langchain/chains/elasticsearch_database/base.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+            from langchain import ElasticsearchDatabaseChain, OpenAI
+            from elasticsearch import Elasticsearch
+
+            db = Elasticsearch("http://localhost:9200")
+            db_chain = ElasticsearchDatabaseChain.from_llm(OpenAI(), db)
+    """
+
+    llm_chain: LLMChain
+    llm: Optional[BaseLanguageModel] = None
```
Comment (baskaryan) [inline langchain/chains/elasticsearch_database/base.py:33]: can delete, was on other class for backwards compatibility

File: langchain/chains/elasticsearch_database/base.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    def _get_indices_infos(self, indices: List[str]) -> List[str]:
+        mappings = self.database.indices.get_mapping(index=",".join(indices))
+        if self.sample_documents_in_index_info > 0:
+            for k, v in mappings.items():
+                hits = self.database.search(index=k, body={"query": {"match_all": {}}}, size=self.sample_documents_in_index_info)['hits']['hits']
+                hits = [str(hit['_source']) for hit in hits]
+                mappings[k]['mappings'] = str(mappings[k]['mappings']) + "\n\n/*\n" + "\n".join(hits) + "\n*/"
+        return "\n\n".join(["Mapping for index {}:\n{}".format(index, mappings[index]['mappings']) for index in mappings])
+
+    # def _sql_search(self, query: str) -> str:
```
Comment (baskaryan) [inline langchain/chains/elasticsearch_database/base.py:91]: lets delete if not needed

File: langchain/chains/elasticsearch_database/base.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    @property
+    def _chain_type(self) -> str:
+        return "elasticsearch_database_chain"
+
+    @classmethod
+    def from_llm(
+        cls,
+        llm: BaseLanguageModel,
+        db: Elasticsearch,
```
Comment (baskaryan) [inline langchain/chains/elasticsearch_database/base.py:167]: should we just call it database for consistency

File: langchain/chains/elasticsearch_database/prompts.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from langchain.output_parsers.list import CommaSeparatedListOutputParser
+from langchain.prompts.prompt import PromptTemplate
+
+
+PROMPT_SUFFIX = """Only use the following Elasticsearch indices:
+{indices_info}
+
+Question: {input}"""
+
+# _DEFAULT_SQL_TEMPLATE = """Given an input question, first create a syntactically correct Elasticsearch SQL query to run, then look at the results of the query and return the answer. Unless the user specifies in his question a specific number of examples he wishes to obtain, always limit your query to at most {top_k} results. You can order the results by a relevant column to return the most interesting examples in the database.
```
Comment (baskaryan) [inline langchain/chains/elasticsearch_database/prompts.py:10]: delete commented out prompts

**Review by baskaryan — COMMENTED** [review]
thanks for pr @samber! a few comments

**Comment by samber** [comment]
Thanks for the review @baskaryan. I just made some fixs.

**Comment by baskaryan** [comment]
cc @jeffvestal

**Comment by baskaryan** [comment]
could we add a notebook demoing the chain to `docs/extras/modules/chains/additional` (they'll be converted to docs)? here's the [SQLDatabaseChain docs](https://python.langchain.com/docs/modules/chains/popular/sqlite) for inspo, but can be much lighter than that. basically the code snippet you have with a blurb at the top about elasticsearch databases

**Comment by samber** [comment]
It must be good now ✅

**Comment by samber** [comment]
Thanks for the review and the awesome work in this project ! 🚀

**Comment by jeffvestal** [comment]
This looks really good, the notebook examples are very helpful. Awesome work!

**Pull request merged**